cmd/car/compile.go

```diff
@@ -0,0 +1,446 @@
+							}
+							// remove from childMap
+							nok := append(otherKids[0:i], otherKids[i+1:]...)
+							childMap[otherCid] = nok
+							break // to next child map entry.
+						}
+					}
+				}
+
+				delete(childMap, origCid)
+			}
+		}
+	}
+
+	if !v2 {
+		// write output
+		outStream := os.Stdout
+		if c.IsSet("output") {
+			outFileName := c.String("output")
+			if outFileName == "" {
+				outFileName = carName
+			}
+			outFile, err := os.Create(outFileName)
+			if err != nil {
+				return err
+			}
+			defer outFile.Close()
+			outStream = outFile
+		}
+
+		if err := carv1.WriteHeader(&carv1.CarHeader{
+			Roots:   roots,
+			Version: 1,
+		}, outStream); err != nil {
+			return err
+		}
+		for c, blk := range outBlocks {
+			if err := util.LdWrite(outStream, c.Bytes(), blk); err != nil {
+				return err
+			}
+		}
+	} else {
+		outFileName := c.String("output")
+		if outFileName == "" {
+			outFileName = carName
+		}
+
+		if outFileName == "-" && !c.IsSet("output") {
+			return fmt.Errorf("cannot stream carv2's to stdout")
+		}
+		bs, err := blockstore.OpenReadWrite(outFileName, roots)
+		if err != nil {
+			return err
+		}
+		for bc, blk := range outBlocks {
+			ob, _ := blocks.NewBlockWithCid(blk, bc)
+			bs.Put(c.Context, ob)
+		}
+		return bs.Finalize()
+	}
+
+	return nil
+}
+
+func serializeBlock(ctx context.Context, codec cid.Prefix, encoding string, raw []byte) (cid.Cid, []byte, error) {
+	ls := cidlink.DefaultLinkSystem()
+	store := memstore.Store{Bag: map[string][]byte{}}
+	ls.SetReadStorage(&store)
+	ls.SetWriteStorage(&store)
+	b := basicnode.Prototype.Any.NewBuilder()
+	if encoding == "json" {
+		if err := dagjson.Decode(b, bytes.NewBuffer(raw)); err != nil {
+			return cid.Undef, nil, err
+		}
+	} else if encoding == "raw" {
+		if err := b.AssignBytes(raw); err != nil {
+			return cid.Undef, nil, err
+		}
+	} else {
+		return cid.Undef, nil, fmt.Errorf("unknown encoding: %s", encoding)
+	}
+	lnk, err := ls.Store(linking.LinkContext{Ctx: ctx}, cidlink.LinkPrototype{Prefix: codec}, b.Build())
+	if err != nil {
+		return cid.Undef, nil, err
+	}
+	outCid := lnk.(cidlink.Link).Cid
+	outBytes, outErr := store.Get(ctx, outCid.KeyString())
+	return outCid, outBytes, outErr
+}
+
+// DebugCar is a command to translate between a car file, and a human-debuggable patch-like format.
+func DebugCar(c *cli.Context) error {
+	var err error
+	inStream := os.Stdin
+	inFile := "-"
+	if c.Args().Len() >= 1 {
+		inFile = c.Args().First()
+		inStream, err = os.Open(inFile)
+		if err != nil {
+			return err
+		}
+	}
+
+	rd, err := carv2.NewBlockReader(inStream)
+	if err != nil {
+		return err
+	}
+
+	// patch the header.
+	outStream := os.Stdout
+	if c.IsSet("output") {
+		outFileName := c.String("output")
+		outFile, err := os.Create(outFileName)
+		if err != nil {
+			return err
+		}
+		defer outFile.Close()
+		outStream = outFile
+	}
+
+	outStream.WriteString("car compile ")
+	if rd.Version == 2 {
+		outStream.WriteString("--v2 ")
```

Comment:
It looks like this is the only hint that it once was a v2 and there's not space to say anything else about it? I guess mostly we care about stability of v1 forms and the v2 is just a convenience wrapper, but if we went ahead with additional features, such as the messaging capability in https://github.com/ipld/go-car/pull/322, then where would we put these things? Would it be hard to extend this format to include those things, and could we do it in a non-breaking way perhaps?

Reply:
this is mirroring [`git patch`](https://www.oreilly.com/library/view/git-pocket-guide/9781449327507/ch11.html)` where the first line mirrors the command used to generate the patch.

Here we use v2 to indicate if the original car was a v2 or not, and can use that as a default to re-build the same car format if it is not specified explicitly on the command line when re-compiling